src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -607,29 +607,29 @@ abstract class BTypes {
    * a missing info. In order not to crash the compiler unnecessarily, the inliner does not force
    * infos using `get`, but it reports inliner warnings for missing infos that prevent inlining.
    */
-  final case class ClassBType(internalName: InternalName)(cache: mutable.Map[InternalName, ClassBType]) extends RefBType {
+  sealed abstract class ClassBType protected(val internalName: InternalName) extends RefBType {
+    def fromSymbol: Boolean
     /**
      * Write-once variable allows initializing a cyclic graph of infos. This is required for
      * nested classes. Example: for the definition `class A { class B }` we have
      *
      *   B.info.nestedInfo.outerClass == A
      *   A.info.nestedClasses contains B
      */
-    private var _info: Either[NoClassBTypeInfo, ClassInfo] = null
+    // volatile is required to ensure no early initialisation in apply
+    // like classic double checked lock in java
+    @volatile private var _info: Either[NoClassBTypeInfo, ClassInfo] = null
 
     def info: Either[NoClassBTypeInfo, ClassInfo] = {
+      if (_info eq null)
+        // synchronization required to ensure the apply is finished
+        // which populates info. ClassBType doesnt escape apart from via the map
+        // and the object mutex is locked prior to insertion. See apply
+        this.synchronized()
```

Comment:
What's the intent of the empty synchronized block here? I'm not sure this should have any effect under the Java Memory Model.

Reply:
it requires the acquire and releaseof the lock - see the comment above

The lock is acquired by apply. I could make this an explicit lock but that would demand another field

Reply:
Thanks for the clarification. I hadn't seen this idiom before :)

Reply:
It does seem however that we could model it better to avoid the synchronization, info and Either/BiasedEither and exception as control flow as something like
```
val classBTypeCache: concurrent.Map[InternalName, ClassBTypeInternal]
private sealed trait ClassBTypeInternal { def resolve: ClassBType}

class ClassBTypePending extends ClassBType {
  val result = Promise[ClassBType ]()
  def resolve = Await.result(result)
}
sealed abstract class ClassBType protected(val internalName: InternalName) extends RefBType with ClassBTypeInternal {
  final def resolve = this
  def get: ClassInfo
  def orThrow: ClassInfo
}
case class ClassInfo extends ClassBType ...
trait NoClassBTypeInfo extends ClassBType 
```

Reply:
@retronym I do think that the empty synchronized block feels dirty though :-(